File: backend/models/Course.js
const mongoose = require('mongoose');

const courseSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Название курса обязательно'],
    trim: true,
    maxlength: [200, 'Название не должно превышать 200 символов']
  },
  description: {
    type: String,
    required: [true, 'Описание курса обязательно'],
    maxlength: [1000, 'Описание не должно превышать 1000 символов']
  },
  content: {
    type: String,
    required: true
  },
  category: {
    type: String,
    enum: ['recycling', 'energy', 'water', 'biodiversity', 'climate', 'general'],
    default: 'general'
  },
  level: {
    type: String,
    enum: ['beginner', 'intermediate', 'advanced'],
    default: 'beginner'
  },
  duration: {
    type: Number, // в минутах
    default: 30,
    min: 1
  },
  points: {
    type: Number, // награда за прохождение
    default: 50,
    min: 0
  },
  thumbnail: {
    type: String
  },
  author: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  modules: [{
    title: {
      type: String,
      required: true
    },
    content: {
      type: String,
      required: true
    },
    order: {
      type: Number,
      required: true
    },
    videoUrl: String,
    resources: [String] // URLs дополнительных материалов
  }],
  quiz: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Quiz'
  },
  published: {
    type: Boolean,
    default: false
  },
  enrolledCount: {
    type: Number,
    default: 0,
    min: 0
  },
  rating: {
    type: Number,
    default: 0,
    min: 0,
    max: 5
  },
  reviews: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    rating: {
      type: Number,
      min: 1,
      max: 5
    },
    comment: String,
    createdAt: {
      type: Date,
      default: Date.now
    }
  }]
}, {
  timestamps: true
});

// Индексы
courseSchema.index({ category: 1, level: 1 });
courseSchema.index({ published: 1, createdAt: -1 });
courseSchema.index({ author: 1 });

// Сортировка модулей по порядку
courseSchema.pre('save', function(next) {
  if (this.modules && this.modules.length > 0) {
    this.modules.sort((a, b) => a.order - b.order);
  }
  next();
});

// Пересчет среднего рейтинга
courseSchema.methods.calculateAverageRating = function() {
  if (this.reviews.length === 0) {
    this.rating = 0;
  } else {
    const sum = this.reviews.reduce((acc, review) => acc + review.rating, 0);
    this.rating = (sum / this.reviews.length).toFixed(1);
  }
  return this.save();
};

module.exports = mongoose.model('Course', courseSchema);

File: frontend/lib/models/course.dart
class Course {
  final String id;
  final String title;
  final String description;
  final String content;
  final String category;
  final String level;
  final int duration; // в минутах
  final int points;
  final String? thumbnail;
  final String authorId;
  final String? authorName;
  final List<CourseModule> modules;
  final String? quizId;
  final bool published;
  final int enrolledCount;
  final double rating;
  final List<CourseReview> reviews;
  final DateTime createdAt;

  Course({
    required this.id,
    required this.title,
    required this.description,
    required this.content,
    required this.category,
    required this.level,
    required this.duration,
    required this.points,
    this.thumbnail,
    required this.authorId,
    this.authorName,
    required this.modules,
    this.quizId,
    required this.published,
    required this.enrolledCount,
    required this.rating,
    required this.reviews,
    required this.createdAt,
  });

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['_id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      content: json['content'] ?? '',
      category: json['category'] ?? '',
      level: json['level'] ?? 'beginner',
      duration: json['duration'] ?? 30,
      points: json['points'] ?? 0,
      thumbnail: json['thumbnail'],
      authorId: json['author'] != null 
          ? (json['author'] is String ? json['author'] : json['author']['_id'] ?? '')
          : '',
      authorName: json['author'] != null && json['author'] is Map 
          ? json['author']['username'] 
          : null,
      modules: (json['modules'] as List?)
          ?.map((m) => CourseModule.fromJson(m))
          .toList() ?? [],
      quizId: json['quiz'],
      published: json['published'] ?? false,
      enrolledCount: json['enrolledCount'] ?? 0,
      rating: (json['rating'] ?? 0).toDouble(),
      reviews: (json['reviews'] as List?)
          ?.map((r) => CourseReview.fromJson(r))
          .toList() ?? [],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
    );
  }
}

class CourseModule {
  final String title;
  final String content;
  final int order;
  final String? videoUrl;
  final List<String> resources;

  CourseModule({
    required this.title,
    required this.content,
    required this.order,
    this.videoUrl,
    required this.resources,
  });

  factory CourseModule.fromJson(Map<String, dynamic> json) {
    return CourseModule(
      title: json['title'],
      content: json['content'],
      order: json['order'],
      videoUrl: json['videoUrl'],
      resources: List<String>.from(json['resources'] ?? []),
    );
  }
}

class CourseReview {
  final String userId;
  final String? username;
  final int rating;
  final String? comment;
  final DateTime createdAt;

  CourseReview({
    required this.userId,
    this.username,
    required this.rating,
    this.comment,
    required this.createdAt,
  });

  factory CourseReview.fromJson(Map<String, dynamic> json) {
    return CourseReview(
      userId: json['user'] != null
          ? (json['user'] is String ? json['user'] : json['user']['_id'] ?? '')
          : '',
      username: json['user'] != null && json['user'] is Map 
          ? json['user']['username'] 
          : null,
      rating: json['rating'] ?? 0,
      comment: json['comment'],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
    );
  }
}

